File: 410/410_version1.py
# Author: Runhua
# Date: 05/20/2019
class Solution:
    def splitArray(self, nums: List[int], m: int) -> int:
        lower_bound,upper_bound = max(nums),sum(nums)+1
        while lower_bound < upper_bound:
            limit = (lower_bound+upper_bound)//2
            curr_group = self.minGroup(nums,limit)
            if curr_group > m:
                lower_bound = limit + 1
            else:
                upper_bound = limit
        return lower_bound
    
    
    def minGroup(self,nums,limit):
        group_num = 1
        curr_sum = 0
        for num in nums:
            if curr_sum+num > limit:
                group_num+=1
                curr_sum = num
            else:
                curr_sum+=num
        return group_num

File: 410/410_version1.java
/**
Author: Runhua
Date: 05/20/2019
*/
class Solution {
    private int[] nums;
    public int splitArray(int[] nums, int m) {
        this.nums = nums;
        long lowerBound = getMax(nums),upperBound = getSum(nums)+1;
        while(lowerBound < upperBound){
            long limit = (lowerBound+upperBound)/2;
            int currGroup = minGroup(limit);
            if(currGroup > m) lowerBound = limit+1;
            else upperBound = limit;
        }
        return (int) lowerBound;
    }
    
    private int minGroup(long limit){
        int currSum = 0,groupNum=1;
        for(int n:nums){
            if(currSum+n > limit){
                groupNum++;
                currSum = n;
            } else{
                currSum+=n;
            }
        }
        return groupNum;
    }
    
    
    private int getMax(int[] nums){
        int result = Integer.MIN_VALUE;
        for(int n:nums) result = Math.max(result,n);
        return result;
    }
    
    private long getSum(int[] nums){
        long result = 0;
        for(int n:nums) result+=n;
        return result;
    }
}
